### PR TITLE
SSO Config: Add generic OAuth

### PR DESCRIPTION
**Pull request by Clarity-89**
**What is this feature?**

Add page for generic OAuth config.

**Why do we need this feature?**

Add provider configuration page for Generic OAuth.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/identity-access-team/issues/522

**Special notes for your reviewer:**

Please check that:
- [X] It works as expected from a user's perspective.
- [X] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by dmihai — COMMENTED** [review]
The page for generic OAuth looks good. I tested the changes locally and found the following minor issues:
- There is no `Enabled` field as we have with the other providers.
- In case there are validation errors in the generic oauth form (for example a team id is not numeric) and the user saves the changes then the browser is being redirected to the main authentication page (the one listing the providers) without showing the actual error to the user. The expected behaviour would be to show the validation errors and don't redirect.
- It seems that `Configure TLS` field doesn't do anything. It is not being sent to the backend on save (expected behaviour) and also it doesn't do anything on frontend.

**Comment by dmihai** [comment]
With the latest fix for not submitting the form in case there are errors, I found the following edge case:
- The user edits the form and then tries to save the changes
- If the form contains validation errors then the validation errors are shown to the user and the form is not being submitted (expected behaviour)
- As soon as the user fixes the errors, the page is being redirected to the main authentication page without submitting the changes to backend. I'd expect here to wait for the user to save the changes before redirecting or submitting the form.

**Comment by dmihai** [comment]
With the latest validation fix I found the following corner case:
- user tries to save a form that contains validation errors
- the validation error is shown and the form is not submitted
- the user fixes the validation error and saves again
- the form is correctly submitted and the browser is redirected back to the main auth page
----- (expected behaviour so far) ------
- the user goes to the generic oauth provider form
- the form contains the old changes (there is a GET request to fetch the provider settings but it seems the "local" settings are rendered instead)
- reloading the page or navigating again to the provider settings seems to fix this

**Comment by Clarity-89** [comment]
> With the latest validation fix I found the following corner case:
> 
> * user tries to save a form that contains validation errors
> * the validation error is shown and the form is not submitted
> * the user fixes the validation error and saves again
> * the form is correctly submitted and the browser is redirected back to the main auth page
>   ----- (expected behaviour so far) ------
> * the user goes to the generic oauth provider form
> * the form contains the old changes (there is a GET request to fetch the provider settings but it seems the "local" settings are rendered instead)
> * reloading the page or navigating again to the provider settings seems to fix this

Pushed a fix for that, a bit hard to test this locally as the save endpoint doesn't seem to be working yet.

**Comment by Clarity-89** [comment]
> It seems that Configure TLS field doesn't do anything. It is not being sent to the backend on save (expected behaviour) and also it doesn't do anything on frontend.

I don't see this field in the backend response, so I guess this is purely for UI purposes? If so I might remove this field altogether.

**Review by dmihai — APPROVED** [review]
I tested the changes locally and it worked, great work!

There is one issue with the client secret. At the moment, the client secret is being sent to the backend even if the user doesn't reset it explicitly. This causes the client secret to be encoded again from the previous version every time the form is saved. The issue will probably get fixed after we merge the latest changes for the GET endpoint.

**Comment by Clarity-89** [comment]
> There is one issue with the client secret. At the moment, the client secret is being sent to the backend even if the user doesn't reset it explicitly. This causes the client secret to be encoded again from the previous version every time the form is saved. The issue will probably get fixed after we merge the latest changes for the GET endpoint.

Hmm yeah, fixing this would be quite tricky on the frontend, so hopefully the backend changes will fix it.

File: public/app/features/auth-config/fields.ts [hunk 1]
```diff
@@ -66,52 +141,194 @@ export const fieldMap: Record<string, FieldData> = {
...
+    label: 'Define Allowed Groups',
+    type: 'switch',
+  },
+  defineAllowedTeamsIds: {
+    label: 'Define Allowed Teams Ids',
+    type: 'switch',
+  },
+  usePkce: {
+    label: 'Use Pkce',
+    description: 'If enabled, it will automatically sync the Grafana server administrator role',
```
Comment (mgyongyosi) [inline public/app/features/auth-config/fields.ts:277]: Can we add links to the description of the label?
```suggestion
    description: 'If enabled, Grafana will use [Proof Key for Code Exchange (PKCE)](https://datatracker.ietf.org/doc/html/rfc7636) with the OAuth2 Authorization Code Grant.',
```

File: public/app/features/auth-config/fields.ts [hunk 2]
```diff
@@ -19,13 +19,81 @@ export const fields: Record<SSOProvider['provider'], Array<keyof SSOProvider['se
...
+      id: 'tls',
+      fields: ['configureTLS', 'tlsSkipVerifyInsecure', 'tlsClientCert', 'tlsClientKey', 'tlsClientCa'],
+    },
+  ],
+};
+
 /**
  * List all the fields that can be used in the form
  */
 export const fieldMap: Record<string, FieldData> = {
```
Comment (mgyongyosi) [inline public/app/features/auth-config/fields.ts:92]: @lwandz13 Could you please review the `description` fields for the generic_oauth settings in this file? For more info you can find more info about the different fields on https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/generic-oauth/#configuration-options

Thanks

File: public/app/features/auth-config/fields.ts [hunk 3]
```diff
@@ -66,52 +141,194 @@ export const fieldMap: Record<string, FieldData> = {
...
+    label: 'Use Pkce',
+    description: 'If enabled, it will automatically sync the Grafana server administrator role',
+    type: 'checkbox',
+  },
+  useRefreshToken: {
+    label: 'Use Refresh Token',
+    description: 'If enabled, it will automatically sync the Grafana server administrator role',
+    type: 'checkbox',
+  },
+  configureTLS: {
```
Comment (mgyongyosi) [inline public/app/features/auth-config/fields.ts:285]: How can we hide a full section and/or different settings based on an expression? For example TLS settings only make sense in on-prem.

Comment (Clarity-89) [reply public/app/features/auth-config/fields.ts:285]: That would require a new property on the field, depends what kinds of conditions we need to support.

File: public/app/features/auth-config/ProviderConfigForm.tsx [hunk 4]
```diff
@@ -74,128 +74,75 @@ export const ProviderConfigForm = ({ config, provider, isLoading }: ProviderConf
...
-                  }}
-                />
-              )}
-            />
+  return (
+    <Page.Contents isLoading={isLoading}>
+      <form onSubmit={handleSubmit(onSubmit)} style={{ maxWidth: '600px' }}>
+        <>
+          <FormPrompt
+            // TODO Figure out why isDirty is not working
```
Comment (mgyongyosi) [inline public/app/features/auth-config/ProviderConfigForm.tsx:82]: Is this still relevant?

File: public/app/features/auth-config/ProviderConfigForm.tsx [hunk 5]
```diff
@@ -74,128 +74,75 @@ export const ProviderConfigForm = ({ config, provider, isLoading }: ProviderConf
...
-              }}
-            />
-            <Field label="Enabled">
-              <Switch {...register('enabled')} id="enabled" label={'Enabled'} />
+              })}
+            </>
+          )}
+          <Box display={'flex'} gap={2} marginTop={6}>
+            <Field>
+              <Button type={'submit'}>{isSaving ? 'Saving...' : 'Save'}</Button>
```
Comment (mgyongyosi) [inline public/app/features/auth-config/ProviderConfigForm.tsx:136]: Could you please also add a `Delete`/`Remove` button with a confirmation modal which says something like `After removing these settings Grafana will use the provider configuration from the system (config file/environment variables) if any.`

Comment (Clarity-89) [reply public/app/features/auth-config/ProviderConfigForm.tsx:165]: Done. It issues a `DELETE` request with config provider name as an id, not sure if any other ID should be used instead.

File: public/app/features/auth-config/fields.ts (hunk 3 above)
Comment (Clarity-89) [reply public/app/features/auth-config/fields.ts]: Added a `hidden` field to both sections and fields and the filtering logic.

File: public/app/features/auth-config/ProviderConfigForm.tsx (hunk 5 above)
Comment (dmihai) [reply public/app/features/auth-config/ProviderConfigForm.tsx:165]: The parameter for `DELETE` is the same with the one we use for `GET`. Not to be confused with the `name` property from the sso settings.

**Review by lwandz13 — APPROVED** [review]

**Review by mgyongyosi — APPROVED** [review]

**Pull request merged**